src/java/org/apache/cassandra/db/guardrails/Threshold.java

```diff
@@ -0,0 +1,282 @@
+            return false;
+
+        Config config = configProvider.apply(state);
+        return value > Math.min(abortValue(config), warnValue(config));
+    }
+
+    /**
+     * Apply the guardrail to the provided value, warning or aborting if appropriate.
+     *
+     * @param value The value to check.
+     * @param what  A string describing what {@code value} is a value of. This is used in the error message if the
+     *              guardrail is triggered. For instance, say the guardrail guards the size of column values, then this
+     *              argument must describe which column of which row is triggering the guardrail for convenience. Note
+     *              that this is only used if the guardrail triggers, so if it is expensive to build, you can put the
+     *              call to this method behind a {@link #triggersOn} call.
+     * @param state The client state, used to skip the check if the query is internal or is done by a superuser.
+     *              A {@code null} value means that the check should be done regardless of the query.
+     */
+    public void guard(long value, String what, @Nullable ClientState state)
+    {
+        if (!enabled(state))
+            return;
+
+        Config config = configProvider.apply(state);
+
+        long abortValue = abortValue(config);
+        if (value > abortValue)
+        {
+            triggerAbort(value, abortValue, what);
+            return;
+        }
+
+        long warnValue = warnValue(config);
+        if (value > warnValue)
+            triggerWarn(value, warnValue, what);
+    }
+
+    private void triggerAbort(long value, long abortValue, String what)
+    {
+        String message = errMsg(false, what, value, abortValue);
+        abort(message);
+    }
+
+    private void triggerWarn(long value, long warnValue, String what)
+    {
+        String message = errMsg(true, what, value, warnValue);
+        warn(message);
+    }
+
+    /**
+     * Creates a new {@link GuardedCounter} guarded by this threshold guardrail.
+     *
+     * @param whatFct A function called when either a warning or abort is triggered by the created counter to
+     *                describe the value. This is equivalent to the {@code what} argument of {@link #guard}, but it is a
+     *                function to allow the output string to be computed lazily (only if a warn/abort ends up being
+     *                triggered).
+     * @param state   The client state, used to skip the check if the query is internal or is done by a superuser.
+     *                A {@code null} value means that the check should be done regardless of the query.
+     * @return the newly created guarded counter
+     */
+    public GuardedCounter newCounter(Supplier<String> whatFct, @Nullable ClientState state)
+    {
+        Threshold threshold = enabled(state) ? this : NEVER_TRIGGERED;
+        return threshold.new GuardedCounter(whatFct, configProvider.apply(state));
+    }
```

Comment:
deadcode; can we remove?

Reply:
Removed, we can add this back when we add the specific guardrails for which this was originally conceived.